docs/design/datacontracts/Thread.md

```diff
@@ -0,0 +1,195 @@
+
+    TS_Hijacked               = 0x00000080,    // Return address has been hijacked
+
+    // unused                 = 0x00000100,
+    TS_Background             = 0x00000200,    // Thread is a background thread
+    TS_Unstarted              = 0x00000400,    // Thread has never been started
+    TS_Dead                   = 0x00000800,    // Thread is dead
+
+    TS_WeOwn                  = 0x00001000,    // Exposed object initiated this thread
+    TS_CoInitialized          = 0x00002000,    // CoInitialize has been called for this thread
+
+    TS_InSTA                  = 0x00004000,    // Thread hosts an STA
+    TS_InMTA                  = 0x00008000,    // Thread is part of the MTA
+
+    // Some bits that only have meaning for reporting the state to clients.
+    TS_ReportDead             = 0x00010000,    // in WaitForOtherThreads()
+    TS_FullyInitialized       = 0x00020000,    // Thread is fully initialized and we are ready to broadcast its existence to external clients
+
+    TS_TaskReset              = 0x00040000,    // The task is reset
+
+    TS_SyncSuspended          = 0x00080000,    // Suspended via WaitSuspendEvent
+    TS_DebugWillSync          = 0x00100000,    // Debugger will wait for this thread to sync
+
+    TS_StackCrawlNeeded       = 0x00200000,    // A stackcrawl is needed on this thread, such as for thread abort
+                                                // See comment for s_pWaitForStackCrawlEvent for reason.
+
+    // unused                 = 0x00400000,
+
+    // unused                 = 0x00800000,
+    TS_TPWorkerThread         = 0x01000000,    // is this a threadpool worker thread?
+
+    TS_Interruptible          = 0x02000000,    // sitting in a Sleep(), Wait(), Join()
+    TS_Interrupted            = 0x04000000,    // was awakened by an interrupt APC. !!! This can be moved to TSNC
+
+    TS_CompletionPortThread   = 0x08000000,    // Completion port thread
+
+    TS_AbortInitiated         = 0x10000000,    // set when abort is begun
+
+    TS_Finalized              = 0x20000000,    // The associated managed Thread object has been finalized.
+                                                // We can clean up the unmanaged part now.
+
+    TS_FailStarted            = 0x40000000,    // The thread fails during startup.
+    TS_Detached               = 0x80000000,    // Thread was detached by DllMain
+}
+
+record struct DacThreadData (
+    uint ThreadId;
+    TargetNUint OsThreadId;
+    ThreadState State;
+    bool PreemptiveGCDisabled
+    TargetPointer AllocContextPointer;
+    TargetPointer AllocContextLimit;
+    TargetPointer Frame;
+    TargetPointer FirstNestedException;
+    TargetPointer TEB;
+    DacGCHandle LastThrownObjectHandle;
+    TargetPointer NextThread;
+);
+```
+
+## Apis of contract
+``` csharp
+DacThreadStoreData GetThreadStoreData();
+DacThreadStoreCounts GetThreadCounts();
```

Comment:
> we constrain the abstractions in our logical layer to either be 1:1 with some runtime type

I would agree with that if the runtime types had a reasonably clean design. The unmanaged Thread in CoreCLR is opposite of that. For historic reasons, Thread contained every thread-local variable ever used by the unmanaged runtime so it has been a dumping ground for a lot of unrelated stuff. Jeremy chipped away on it a little bit in https://github.com/dotnet/runtime/pull/99989. 

I think we need to some plan to clean this up so that we do not end up faulting a ton of cruft into the contracts.

One possible way to do that is to make some of the contracts identical between CoreCLR, NativeAOT and possibly Mono and refactor the code accordingly. Contracts that are identical between CoreCLR, NativeAOT and possibly Mono have a high chance of being properly factored.

Reply:
> I think we need to some plan to clean this up so that we do not end up faulting a ton of cruft into the contracts.

A general issue here is describing what we want with what we have. This model creates a fair bit of friction with by codifying current asperations instead of reality and permitted evolution with where we want to go. Either we should focus on the asperation and move in that direction or accept the current reality and model it as such while letting the versioning scheme grow as we improve.

Reply:
I think the data contract project is going to fail if we accept the current reality and try to create contracts from it. We do not need to cleanup every little detail, but I think fair bit of cleanup and refactoring is required for the data contract project to succeed.

Reply:
> I think the data contract project is going to fail if we accept the current reality and try to create contracts from it.

I disagree with that assessment, but I accept I am missing some nuance of this approach that forces a cleaner implementation.

> We do not need to cleanup every little detail, but I think fair bit of cleanup and refactoring is required for the data contract project to succeed.

I see. It sounds like there is a prep work then with this effort. Has that specifically been identified and accounted for in the work for the cDAC? As in do we have a "simplication" work item for specific areas of the runtime we expect to focus on for a .NET 9 deliverable? I think it is fine if it is implied in the work item, but it should be an explicit expectation somewhere.

Reply:
> I would agree with that if the runtime types had a reasonably clean design

I wasn't claiming that we are stuck with the shape of the runtime as it exists today. If we want to change the runtime during .NET 9, then ship a contract that is 1:1 with that new runtime shape that exists when we ship .NET 9 that is fine by me.